### PR TITLE
Media Type Adapters

### LINKED ISSUE #5: Feature Request: Preprocessing Transform
# Feature Request: Preprocessor Step

Most of the time, users are working with text in a handful of known media types (markdown, html, etc) or images. 

This proposal is to add a generic preprocessing step on collection creation for preprocessing media into vectors so users don't have to operate on vectors directly

Something similar to 
```python
docs: Collection =vx.create_collection(
    docs='docs',
    dimension=512,
    transform = TextPreprocessor(          # this is new
        chunker= markdown_chunker,
        model= "sentence-transformers/all-Mini-L6-v2"
    )    
)

docs.upsert([
    ("id_0", "# Some markdown", {}),
    ("id_1", "# Some markdown", {}),
])
```

Open questions:
- Should chunking be supported? If so, separately or with the preprocessor?
- Which media types are required?

### PR DESCRIPTION
**Pull request by olirice**
## What kind of change does this PR introduce?
Introduces an adapter system enabling users to interact with collections using text/images/generic media rather than vectors.

### Goals
- Users do not have to adopt the adapter system (BYO-vectors is supported)
- When using the adapter system, users should never see a vector
- `Adapters` are executed lazily to enable large bulk inserts without memory issues 
- `dimension` does not need to be provided when using an adapter
- heavy weight adapter dependencies are isolated behind feature flag
e.g.
```sh
pip install "vecs[text_embedding]"
```

Example of expected flow:

```python
import vecs
from vecs.adapter import Adapter
from vecs.adapter.text import ParagraphChunker, TextEmbedding

# create vector store client
vx = vecs.Client('postgresql://postgres:password@localhost:5611/vecs_db')

# create a collection named docs with 3 dimensional vectors
docs = vx.get_or_create_collection(
    name="docs",
    adapter=Adapter(
        [
            ParagraphChunker(),
            TextEmbedding(model='all-Mini-LM6-v2'),
        ]
    )
)

# add records to the collection using text as the media type
docs.upsert(
    records=[
        (
         "vec0", 
         "four score and ....", 
         {"year": 1973}
        ),
        (
         "vec1", 
         "hello, world!",
         {"year": "2012"}
        )
    ]
)

# search by text
docs.query(data="foo bar")
```

Tasks:
- [x] add `get_or_create_collection` for idempotent scripts 
- [x] deprecated `get_collection` and `create_collection`
- [x] tests that deprecated methods produce warnings
- [x] update documentation to use `get_or_create_collection`
- [x] generic `Adapter` system of composable `AdapterStep`s
- [x] existing tests pass (backwards compatibility)
- [x] tests for generic adapters 
- [x] create a `ParagraphChunking` adapter
- [x] create a `TextEmbedding` adapter
- [x] document implemented adapters
- [x] integration test for example adapters
- [x] document adapter concept
- [x] docstrings everywhere

---
Test Coverage Report

```
============================= test session starts ==============================
platform darwin -- Python 3.9.9, pytest-7.4.0, pluggy-1.2.0
rootdir: /Users/oliverrice/Documents/supabase/vecs
configfile: pytest.ini
plugins: cov-4.1.0, anyio-3.7.1
collected 28 items

src/tests/test_adapters.py ........                                      [ 28%]
src/tests/test_client.py .....                                           [ 46%]
src/tests/test_collection.py ...............                             [100%]

---------- coverage: platform darwin, python 3.9.9-final-0 -----------
Name                           Stmts   Miss  Cover
--------------------------------------------------
src/vecs/__init__.py               8      0   100%
src/vecs/adapter/__init__.py       4      0   100%
src/vecs/adapter/base.py          30      0   100%
src/vecs/adapter/noop.py          11      0   100%
src/vecs/adapter/text.py          33      0   100%
src/vecs/client.py                50      0   100%
src/vecs/collection.py           256      0   100%
src/vecs/exc.py                   16      0   100%
--------------------------------------------------
TOTAL                            408      0   100%


============================== 28 passed in 5.06s ==============================
```
---

resolves #5

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -187,3 +176,96 @@ with vecs.create_client(DB_CONNECTION) as vx:
...
+
+### Create collection
+
+!!! note
+    Deprecated: use [get_or_create_collection](#get-or-create-a-collection)
+
+You can create a collection to store vectors specifying the collections name and the number of dimensions in the vectors you intend to store.
+
+```python
+docs = vx.get_or_create_collection(name="docs", dimension=3)
```
Comment (gregnr) [inline docs/api.md:256]: Should this be `create_collection()`?

```suggestion
docs = vx.create_collection(name="docs", dimension=3)
```

File: docs/concepts_adapters.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
+# Adapters
+
+Adapters are an optional feature to transform data before adding to or querying from a collection. Adapters provide a customizable and modular way to express data transformations and make interacting with collections more erganomic.
```
Comment (gregnr) [inline docs/concepts_adapters.md:3]:
```suggestion
Adapters are an optional feature to transform data before adding to or querying from a collection. Adapters provide a customizable and modular way to express data transformations and make interacting with collections more ergonomic.
```

File: docs/concepts_adapters.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+- [ParagraphChunker](#paragraphchunker)
+- [TextEmbedding](#textembedding)
+
+Have an idea for a useful adapter? [Open an issue](https://github.com/supabase/vecs/issues/new/choose) requesting it.
+
+### ParagraphChunker
+
+The `ParagraphChunker` `AdapterStep` splits text media into paragraphs and yields each paragraph as a separate record. That can be a useful pre-processing step when upserting large documents that contain multiple paragraphs. The `ParagraphChunker` delimits paragraphs by two consecutive line breaks `\n\n`.
+
+`ParagrphChunker` is a pre-preocessing step and must be used in combination with another adapter step like `TextEmbedding` to transform the chunked text into a vector.
```
Comment (gregnr) [inline docs/concepts_adapters.md:80]:
```suggestion
`ParagrphChunker` is a pre-preprocessing step and must be used in combination with another adapter step like `TextEmbedding` to transform the chunked text into a vector.
```

File: docs/concepts_adapters.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    )
+)
+```
+
+When querying the collection, you probably do not want to chunk the text. To skip text chunking during queries, set the `skip_during_query` argument to `True`. Setting `skip_during_query` to `False` will raise an exception if the input text contains more than one paragraph.
+
+
+### TextEmbedding
+
+The `TextEmbedding` `AdapterStep` accepts text and converts it into a vector that can be consumed by the `Collection`. `TextEmbedding` supports all models available in the `sentence_transformers` package. A complete list of supported models is available in `vecs.adapter.TextEmbeddingModel`.
```
Comment (gregnr) [inline docs/concepts_adapters.md:104]:
```suggestion
The `TextEmbedding` `AdapterStep` accepts text and converts it into a vector that can be consumed by the `Collection`. `TextEmbedding` supports all models available in the [`sentence_transformers`](https://www.sbert.net) package. A complete list of supported models is available in `vecs.adapter.TextEmbeddingModel`.
```

File: docs/concepts_adapters.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+# search by text
+docs.query(data="foo bar")
+```
+
+## Interface
+
+Adapters are objects that take in data in the form of `Iterable[Tuple[str, Any, Optional[Dict]]]` where `Tuple[str, Any, Optional[Dict]]]` represents records of `(id, media, metadata)`.
+
+The main use of Adapters is to transform the media part of the records into a form that is ready to be ingested into the collection (like converting text into embeddings). However, Adapters can also modify the id or metadata if required.
```
Comment (gregnr) [inline docs/concepts_adapters.md:127]:
```suggestion
The main use of Adapters is to transform the media part of the records into a form that is ready to be ingested into the collection (like converting text into embeddings). However, Adapters can also modify the `id` or `metadata` if required.
```

**Review by gregnr — APPROVED** [review]

File: docs/api.md (hunk 1 above)
Comment (olirice) [reply docs/api.md:256]: yes, thanks! find/replace error

File: docs/concepts_adapters.md (hunk 3 above)
Comment (olirice) [reply docs/concepts_adapters.md:80]: pushed fix in separate commit

**Pull request merged**